tetragonpod/controller/tetragonpod_controller.go

```diff
@@ -0,0 +1,194 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"context"
+
+	ciliumiov1alpha1 "github.com/cilium/tetragon/tetragonpod/api/v1alpha1"
+	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+// TetragonPodReconciler reconciles a TetragonPod object
+type TetragonPodReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+}
+
+//+kubebuilder:rbac:groups=cilium.io,resources=TetragonPods,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=cilium.io,resources=TetragonPods/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=cilium.io,resources=TetragonPods/finalizers,verbs=update
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+
+// Reconcile is the main loop of the controller that something changes with the pod resources in the cluster.
+func (r *TetragonPodReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	l := log.FromContext(ctx)
+
+	// Get the Pod.
+	pod := &corev1.Pod{}
+	if err := r.Get(ctx, req.NamespacedName, pod); err != nil {
+
+		// CASE 1: Pod is deleted, therefore delete the corresponding tetragonPod resource.
+		if client.IgnoreNotFound(err) == nil {
+			l.Info("Pod Deleted: ", "Name", req.Name, " Namespace", req.Namespace)
+			deleteTetragonPod(ctx, r, req, l)
+			return ctrl.Result{}, nil
+		}
+	}
+
+	// CASE 2: Pod is created or updated.
+	// wait until the pod is running for cleaner logs.
+	if pod.Status.Phase != corev1.PodRunning {
+		return ctrl.Result{}, nil
+	}
+
+	// Check if corresponsind tetragonPod already exists.
+	tetragonPod, exists := checkIfAlreadyExists(ctx, req, r, l)
+	if exists {
+		// CASE 2.1: TetragonPod already exists, update it.
+		l.Info("Pod Updated: ", "Name", req.Name, "Namespace", req.Namespace)
+		updateTetragonPod(ctx, r, l, pod, tetragonPod)
+		return ctrl.Result{}, nil
+	}
+
+	// CASE 2.2: TetragonPod does not exist, create new.
+	l.Info("Pod Created: ", "Name", req.Name, "Namespace", req.Namespace)
+	createTetragonPod(ctx, pod, r, l)
+	return ctrl.Result{}, nil
+}
+
+// deleteTetragonPod deletes the corresponding tetragonPod resource.
+func deleteTetragonPod(ctx context.Context, r *TetragonPodReconciler, req ctrl.Request, l logr.Logger) (ctrl.Result, error) {
```

Comment:
```suggestion
func (r *TetragonPodReconciler) deleteTetragonPod(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
```
then just call `log.FromContext(ctx)` to access the logger. same for other functions.

Reply:
done